server/src/main/java/com/cloud/network/NetworkModelImpl.java

```diff
@@ -1703,6 +1690,109 @@ public void checkNetworkPermissions(Account owner, Network network) {
+                throw new PermissionDeniedException(String.format("network %s cannot be operated by domain admin %s", network, owner));
+            }
+        } else if (owner.getType() == Account.ACCOUNT_TYPE_NORMAL) {
+            if (owner.getType() != Account.ACCOUNT_TYPE_PROJECT && networkOwner.getType() == Account.ACCOUNT_TYPE_PROJECT) {
+                checkProjectNetworkPermissions(owner, networkOwner, network);
+            } else if (networkOwner.getAccountId() != owner.getAccountId()) {
+                throw new PermissionDeniedException(String.format("network %s cannot be operated by normal user %s", network, owner));
+            }
+        } else {
+            throw new PermissionDeniedException(String.format("network %s cannot be operated by this account %s", network, owner));
+        }
+    }
+
+    private void checkSharedNetworkOperatePermissions(Account owner, Network network) {
+        NetworkOffering networkOffering = _networkOfferingDao.findById(network.getNetworkOfferingId());
+        if (networkOffering.isSpecifyVlan() && owner.getType() != Account.ACCOUNT_TYPE_ADMIN) {
+            throw new PermissionDeniedException(String.format("Shared network %s with specifyvlan=true can only be operated by root admin", network));
+        }
+        if (owner.getType() == Account.ACCOUNT_TYPE_DOMAIN_ADMIN) {
+            if (network.getAclType() == ACLType.Domain) {
+                // Allow domain admins to operate shared network for their domain.
+                NetworkDomainVO networkDomainMap = _networkDomainDao.getDomainNetworkMapByNetworkId(network.getId());
+                if (networkDomainMap == null) {
+                    throw new CloudRuntimeException(String.format("Cannot find domain info for Shared network %s with aclType=Domain", network));
+                }
+                if (!_domainDao.isChildDomain(owner.getDomainId(), networkDomainMap.getDomainId())) {
+                    throw new PermissionDeniedException(String.format("Shared network %s cannot be operated by domain admin %s", network, owner));
+                }
+            } else if (network.getAclType() == ACLType.Account) {
+                // Allow domain admins to operate shared network for an account in their domain.
+                NetworkAccountVO networkAccountMap = _networkAccountDao.getAccountNetworkMapByNetworkId(network.getId());
+                if (networkAccountMap == null) {
+                    throw new CloudRuntimeException(String.format("Cannot find account info for Shared network %s with aclType=Account", network));
+                }
+                if (!_domainDao.isChildDomain(owner.getDomainId(), _accountDao.findByIdIncludingRemoved(networkAccountMap.getAccountId()).getDomainId())) {
+                    throw new PermissionDeniedException(String.format("Shared network %s cannot be operated by domain admin %s", network, owner));
```

Comment:
this has the exact same message text as line https://github.com/apache/cloudstack/pull/5730/files#diff-05d77af980dadebef650b49cb8ab34f4d0040fa18ec7643cd2c945de5227933dR1759
maybe externalise or include the reason for throwing in the text so it can be identified more easily?